src/rpp/rpp/operators/on_error_resume_next.hpp

```diff
@@ -17,55 +17,100 @@
 
 namespace rpp::operators::details
 {
+    template<rpp::constraint::observer TObserver>
+    struct on_error_resume_next_disposable final : public rpp::composite_disposable
+    {
+        on_error_resume_next_disposable(TObserver&& observer)
+            : rpp::composite_disposable{}
+            , observer(std::move(observer))
+        {
+        }
+
+        RPP_NO_UNIQUE_ADDRESS TObserver observer;
+    };
+
+    template<rpp::constraint::observer TObserver>
+    struct on_error_resume_next_inner_observer_strategy
+    {
+        using preferred_disposable_strategy = rpp::details::observers::none_disposable_strategy;
+
+        std::shared_ptr<TObserver> observer;
+
+        template<typename T>
+        void on_next(T&& v) const
+        {
+            observer->on_next(std::forward<T>(v));
+        }
+
+        void on_error(const std::exception_ptr& err) const
+        {
+            observer->on_error(err);
+        }
+
+        void on_completed() const
+        {
+            observer->on_completed();
+        }
+
+        void set_upstream(const disposable_wrapper& d) { observer->set_upstream(d); }
+
+        bool is_disposed() const { return observer->is_disposed(); }
+    };
+
+
     template<rpp::constraint::observer TObserver, rpp::constraint::decayed_type Selector>
     struct on_error_resume_next_observer_strategy
     {
         using preferred_disposable_strategy = rpp::details::observers::none_disposable_strategy;
 
-        RPP_NO_UNIQUE_ADDRESS mutable TObserver observer;
-        RPP_NO_UNIQUE_ADDRESS Selector          selector;
-        // Manually control disposable to ensure observer is not used after move in on_error emission
-        mutable rpp::composite_disposable_wrapper disposable = composite_disposable_wrapper::make();
+        on_error_resume_next_observer_strategy(TObserver&& observer, const Selector& selector)
+            : state{init_state(std::move(observer))}
+            , selector{selector}
+        {
+        }
 
-        RPP_CALL_DURING_CONSTRUCTION(
-            observer.set_upstream(disposable););
+        std::shared_ptr<on_error_resume_next_disposable<TObserver>> state;
+        RPP_NO_UNIQUE_ADDRESS Selector                              selector;
 
         template<typename T>
         void on_next(T&& v) const
         {
-            observer.on_next(std::forward<T>(v));
+            state->observer.on_next(std::forward<T>(v));
         }
 
         void on_error(const std::exception_ptr& err) const
         {
-            std::optional<std::invoke_result_t<Selector, std::exception_ptr>> selector_obs;
             try
             {
-                selector_obs.emplace(selector(err));
+                selector(err).subscribe(on_error_resume_next_inner_observer_strategy<TObserver>{std::shared_ptr<TObserver>(state, &state->observer)});
```

Comment:
Do we need the drain approach for this operator too?

Reply:
Not sure because it applies only once